aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -1263,6 +1290,123 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
   }
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ eig ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+static void apply_eig(const Tensor& self, bool eigenvectors, Tensor& out_eigvals, Tensor& out_eigvecs,
+                      int64_t *info_ptr) {
+#ifndef USE_MAGMA
+AT_ERROR("symeig: MAGMA library not found in "
+    "compilation. Please rebuild with MAGMA.");
+#else
+  magma_vec_t jobvr = eigenvectors ? MagmaVec : MagmaNoVec;
+  magma_int_t n = magma_int_cast(self.size(-1), "n");
+  auto self_data = self.data_ptr<scalar_t>();
+
+  out_eigvals.resize_({2, n});
+  auto out_eigvals_data = out_eigvals.data_ptr<scalar_t>();
+  scalar_t *wr = out_eigvals_data;
+  scalar_t *wi = out_eigvals_data+n;
+
+  scalar_t *vr_data = NULL;
+  magma_int_t ldvr = 1;
+  if (jobvr == MagmaVec)
+  {
+      vr_data = out_eigvecs.data_ptr<scalar_t>();
+      ldvr = n;
+  }
+
+  if (n > 0) {
+    // call magmaEig once to get the optimal size of work_data
+    scalar_t wkopt;
+    magma_int_t info;
+    magmaEig<scalar_t>(MagmaNoVec, jobvr, n, self_data, n, wr, wi, NULL, 1, vr_data, ldvr, &wkopt, -1, &info);
+    magma_int_t lwork = (magma_int_t) wkopt;
+
+    // call it a 2nd time to to the actual work
+    scalar_t *work_data = nullptr;
+    ALLOCATE_ARRAY(work_data, scalar_t, lwork);
+    magmaEig<scalar_t>(MagmaNoVec, jobvr, n, self_data, n, wr, wi, NULL, 1, vr_data, ldvr, work_data, lwork, &info);
+    *info_ptr = info;
+  }
+
+  // magmaEig returns column-ordered tensors, turn them into row-ordered
+  out_eigvals.transpose_(0, 1);
+  if (jobvr == MagmaVec)
+      out_eigvecs.transpose_(0, 1);
+#endif
+}
+
+/*
+ * Internal helper; like eig_cuda but:
+ *   1. assume that self is a square matrix of side "n"
+ *   2. return CPU tensors, which will be copied to GPU memory by the caller
+ */
+static std::tuple<Tensor,Tensor> eig_cuda_helper(const Tensor & self, int64_t n, bool eigenvectors) {
+  // copy self to pinned CPU memory
+  auto self_working_copy = at::empty(self.sizes(),
+                                     at::TensorOptions(at::kCPU).dtype(self.dtype()).pinned_memory(true));
+
+  self_working_copy.transpose_(0, 1); // make it column-ordered, what magmaEig wants
+  self_working_copy.copy_(self);
```

Comment:
The purpose of `transpose_` here is to fix the strides to make `self_working_copy` column-ordered tensor. As a side-effect, its data content of undefined values is changed as well. This is unnecessary.

Suggestion: use `auto self_working_copy = at::empty_strided(sizes, strides);` where `sizes == self.sizes()` and `strides=reversed(self.strides())` so that the inplace transpose can be eliminated.

Reply:
good point, I wasn't aware of `at::empty_strided`. Fixed in 3ee78b0